lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,377 @@
+        return RTA.trace_events
+
+    @property
+    @memoized
+    def fast_ramp(self):
+        # The EWMA "fast ramp up" feature has been merged only in kernel TBD
+        # TODO fix kernel version
+        min_kernel = KernelVersion('4.5-rc1').parts
+
+        cur_kernel = self.plat_info['kernel']['version'].parts
+        return cur_kernel < min_kernel
+
+    def _plot_signals(self, df, failures):
+        fig, axes = plt.subplots(1, 1, figsize=(16, 8))
+        for start in failures:
+            print(start)
+            axes.axvline(start, alpha=0.5, color='r')
+        df[['util_avg', 'util_est_enqueued', 'util_est_ewma']]\
+            .plot(ax=axes, drawstyle='steps-post')
+        figname = os.path.join(self.res_dir, 'util_est_signals.png')
+        pl.savefig(figname, bbox_inches='tight')
+        plt.close()
+
+    @requires_events('sched_util_est_task', 'sched_load_se')
+    @RTAEventsAnalysis.task_phase_windows.used_events
+    def test_areas(self) -> ResultBundle:
+        """
+        Test signals are properly "dominated".
+
+        The integral of `util_est_enqueued` is expected to be always not
+        smaller than that of `util_avg`, since this last is subject to decays
+        while the first not.
+
+        On fast-ramp systems, the `util_est_ewma` signal is never smaller then
+        the `util_est_enqueued`, thus his integral is expected to be bigger.
+
+        On non fast-ramp systems instead, the `util_est_ewma` is expected to be
+        smaller then `util_est_enqueued` in ramp-up phases, or bigger in
+        ramp-down phases.
+
+        Those conditions are checked on a single execution of a task which has
+        three main behaviours:
+        - STABLE: periodic big task running for a relatively long period to
+                  ensure `util_avg` saturation.
+        - DOWN: periodic ramp-down task, to slowly decay `util_avg`
+        - UP: periodic ramp-up task, to slowly increase `util_avg`
+
+        """
+        PhaseStats = namedtuple("PhaseStats", [
+            'start', 'end', 'area_util', 'area_enqueued', 'area_ewma'])
+        failure_reasons = {}
+        metrics = {}
+
+        # We have only two task: the main 'rt-app' task and our 'test_task'
+        test_task = self.trace.analysis.rta.rtapp_tasks[-1]
+
+        ue_df = self.trace.df_events('sched_util_est_task')
+        ue_df = ue_df[ue_df.pid == test_task.pid]
+        ua_df = self.trace.df_events('sched_load_se')
+        ua_df = ua_df[ua_df.pid == test_task.pid]
+
+        passed = True
+        for phase in self.trace.analysis.rta.task_phase_windows(test_task):
+            phase_df = ue_df[phase.start:phase.end]
+            area_enqueued = series_integrate(phase_df.util_est_enqueued)
+            area_ewma = series_integrate(phase_df.util_est_ewma)
+
+            phase_df = ua_df[phase.start:phase.end]
+            area_util = series_integrate(phase_df.util)
+
+            metrics[phase.id] = PhaseStats(phase.start, phase.end,
```

Comment:
Since metrics are used by external entities, it's better to use nested dict than adhoc type for them. Otherwise, whatever is behind the type becomes opaque to these external data processors and we also will not be able to format them nicely in test output (once we get there).

Reply:
Does that means I should drop the namedtuple?!?
Because they definitively read better in my notebook... perhaps we should make the "external entities" smarter... Or, can't we add a dict-ionification method to the namedtuple, to be used by the external dummy entity?

Reply:
I would suggest dropping them for the metrics indeed, or indeed we can try to make them instances of `collections.abc.Mapping`. I can have a go at that after we merge it though